README.rst

```diff
@@ -8,5 +8,195 @@
+built upon this package my provide them). These include, but are not
+limited to:
+
+- Providing a user interface for managing subscriptions.
+- Providing an HTTPS API for managing subscriptions. This package
+  provides the underlying data storage, but accepting parameters, etc,
+  and marshaling them into the correct Python calls, is not a concern
+  here.
+- Providing a user interface or HTTPS API for viewing webhook audit
+  logs.
+
+In Scope/Features
+=================
+
+Certain concerns are very much in scope for this package, and this
+package should provide a complete, easy to use solution that addresses
+these concerns. Where necessary, if a concern cannot be addressed
+directly by this package, extension points (interfaces and
+``zope.component`` utilities) may be defined. These include, but are
+not limited to:
+
+- Resource Representation
+
+  The on-the-wire form of the resources is built using
+  `nti.externalization <https://ntiexternalization.readthedocs.io>`_.
+
+  To allow customization of the external forms, a named externalizer
+  is used; nti.externalization will fall back to the default
+  externalizer if no externalizer of the given name is available. The
+  default externalizer is named "webhook-delivery", but dialects may
+  use something different.
+
+- Alternate Webhook Dialects
+
+  Webhooks are a general protocol and mostly interoperable. But to
+  support cases where particular destinations have specific
+  requirements, "dialects" are used. There is a default dialect and
+  then there may be specializations of it. Each webhook subscription
+  may have associated with it the name of a dialect to use. These
+  dialects are found in the component registry. For example, a dialect
+  may choose to use a different externalizer name such as
+  "zapier-webhook-delivery".
+
+- Transactional
+
+  Webhooks should not be delivered if the ultimate creation or
+  persistence of a resource failed. To this end, webhook delivery in
+  this package is integrated with the `transaction
+  <https://transaction.readthedocs.io>`_ package.
+
+  Resources are externalized during a late phase of the transaction
+  commit process; the details about the delivery are recorded and
+  persisted, and only after the transaction is successfully committed
+  does the HTTP request get made.
+
+- Concurrency
+
+  Webhook delivery and record keeping should be lightweight, and
+  all actual network IO should proceed in a non-blocking fashion. This
+  means that this package will spawn threads (or greenlets, using
+  `gevent <http://www.gevent.org>`_.
+
+- Error Handling/Failure Retry
+
+  A limited amount of retry logic is provided by this package, but
+  that does not extend to process boundaries. If the process hosting
+  this package is killed while a delivery is pending, no automatic
+  provision is made to resume delivery attempts in any other process.
+
+  The API is present to allow that to be implemented, though.
```

Comment:
I think this is probably ok. Essentially the containing app server would need to have logic to restart the retry process on startup (if necessary)? I think provided we can envision how we solve that in our application stack that I'm good with this.

Reply:
It may not be fast, and it may actually need to be a separate command launched manually or scheduled in order to preserve at-most-once delivery (we don't want to get into a situation where multiple processes try to deliver the same payload for the same event; assuming one event is in one process simplifies things somewhat), but I can imagine it would be solvable if needed.